File: ios/Classes/Pastie.m
//
//  Pastie.m
//  Paste
//
//  Created by Grant Paul on 7/5/10.
//  Copyright 2010 Xuzz Productions, LLC. All rights reserved.
//

#import "Pastie.h"

#define kHeaderBoundary @"_xuzz_productions_paste_"

@implementation Pastie
@synthesize delegate;

+ (NSDictionary *)languages {
    return [NSDictionary dictionaryWithObjectsAndKeys:
            [NSNumber numberWithInt:2],  @"ActionScript",
            [NSNumber numberWithInt:13], @"Bash (shell)",
            [NSNumber numberWithInt:20], @"C#",
            [NSNumber numberWithInt:7],  @"C/C++",
            [NSNumber numberWithInt:8],  @"CSS",
            [NSNumber numberWithInt:5],  @"Diff",
            [NSNumber numberWithInt:21], @"Go",
            [NSNumber numberWithInt:12], @"HTML (ERB / Rails)",
            [NSNumber numberWithInt:11], @"HTML / XML",
            [NSNumber numberWithInt:9],  @"Java",
            [NSNumber numberWithInt:10], @"JavaScript",
            [NSNumber numberWithInt:1],  @"Objective-C/C++",
            [NSNumber numberWithInt:18], @"Perl",
            [NSNumber numberWithInt:15], @"PHP",
            [NSNumber numberWithInt:6],  @"Plain Text",
            [NSNumber numberWithInt:16], @"Python", 
            [NSNumber numberWithInt:3],  @"Ruby", 
            [NSNumber numberWithInt:4],  @"Ruby on Rails", 
            [NSNumber numberWithInt:14], @"SQL", 
            [NSNumber numberWithInt:19], @"YAML",  
    nil];
}

- (void)beginSubmissionWithText:(NSString *)text makePrivate:(BOOL)private language:(NSInteger)language {
	NSMutableDictionary *post_dict = [NSMutableDictionary dictionary];
	[post_dict setObject:text forKey:@"paste[body]"];
	[post_dict setObject:@"burger" forKey:@"paste[authorization]"];
	[post_dict setObject:private ? @"1" : @"0" forKey:@"paste[restricted]"];
	[post_dict setObject:[NSString stringWithFormat:@"%d", language] forKey:@"paste[parser_id]"];
	
	NSMutableData *post_data = [NSMutableData data];
	for (NSString *key in [post_dict allKeys]) {
		[post_data appendData:[[NSString stringWithFormat:@"--%@\r\n", kHeaderBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
		[post_data appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n\r\n", key] dataUsingEncoding:NSUTF8StringEncoding]];
		[post_data appendData:[[post_dict valueForKey:key] dataUsingEncoding:NSUTF8StringEncoding]];
		[post_data appendData:[[NSString stringWithString:@"\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
	}
	[post_data appendData:[[NSString stringWithFormat:@"--%@--\r\n", kHeaderBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
	
	
	NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"http://pastie.org/pastes"]];
	NSString *content_type = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", kHeaderBoundary];
	[request addValue:content_type forHTTPHeaderField:@"Content-Type"];
	[request setHTTPMethod:@"POST"];
	[request setHTTPBody:post_data];
	
	[NSURLConnection connectionWithRequest:request delegate:self];
}

- (void)beginSubmissionWithText:(NSString *)text makePrivate:(BOOL)private {
    [self beginSubmissionWithText:text makePrivate:private language:[[[[self class] languages] objectForKey:@"Plain Text"] intValue]];
}

- (void)beginSubmissionWithText:(NSString *)text {
    [self beginSubmissionWithText:text makePrivate:NO];
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
	if ([delegate respondsToSelector:@selector(submissionFailedWithError:)])
		[delegate submissionFailedWithError:nil];
}

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
	if ([delegate respondsToSelector:@selector(submissionCompletedWithURL:)])
		[delegate submissionCompletedWithURL:[response URL]];
}

@end
 

File: webos/enyo/Pastie.js
enyo.kind({
    name: "Pastie",
    kind: enyo.Object,

    statics: {
        _languages: {
            "ActionScript":2,
            "Bash (shell)":13,
            "C#":20,
            "C/C++":7,
            "CSS":8,
            "Diff":5,
            "Go":21,
            "HTML (ERB / Rails)":12,
            "HTML / XML":11,
            "Java":9,
            "JavaScript":10,
            "Objective-C/C++":1,
            "Perl":18,
            "PHP":15,
            "Plain Text":6,
            "Python":16,
            "Ruby":3,
            "Ruby on Rails":4,
            "SQL":14,
            "YAML":19
        },
        languages: function() {
            var names = [];
            for (var name in this._languages) {
                if (this._languages.hasOwnProperty(name))
                    names.push(name);
            }
            return names;
        },
        languageNamed: function(name) {
            return this._languages[name];
        }
    },

    published: {
        language: undefined,
        text: "",
        privately: false
    },
    constructor: function() {
        this.inherited(arguments);
    },

    submit: function(finished, failed) {
	    var boundary = "_xuzz_productions_paste_";
	    var body = "";
	    var post = {
		    "paste[body]": this.text,
		    "paste[authorization]": "burger",
		    "paste[restricted]": this.privately ? "1" : "0",
		    "paste[parser_id]": Pastie.languageNamed(this.language)
	    };
	    
        for (var key in post) {
		    body += "--" + boundary + "\r\n";
		    body += "Content-Disposition: form-data; name=\"" + key + "\"\r\n\r\n";
		    body += post[key] + "\r\n";
	    }

	    body += "--" + boundary + "--\r\n";
    
	    var http = new XMLHttpRequest();
	    http.open("POST", "http://xuzz.net/apps/paste.php", true); //"http://pastie.org/pastes", true);
	    http.setRequestHeader("Content-Type", "multipart/form-data; charset=UTF-8; boundary=" + boundary);
	    http.setRequestHeader("Content-Length", body.length);
	    
        http.onreadystatechange = function() {
		    if (http.readyState == 4) {
				var response = http.responseText;

			    if ((http.status == 200 || http.status == 0) && response != "") {
				    // var url = response.match(/http:\/\/pastie\.org\/pastes\/(\d*)\/text/g);
			        finished(response);
                } else {
			        failed(http.status);
                }
		    }
	    }
    
	    http.send(body);
    }
});
